### DISCUSSION TITLE
Error: [redi]: Cannot find "univer.doc.text-selection-render-manager" registered by any injector.

### DISCUSSION THREAD

**Discussion by Maxth47** [dream-num/univer#2863] [Q&A | 常见问题] [answered]
I get an error like this in my React app using Typescript:
`Error: [redi]: Cannot find "univer.doc.text-selection-render-manager" registered by any injector. The stack of dependencies is: "univer.doc.text-selection-render-manager".`

I found the fix for the issue [here](https://github.com/dream-num/univer/discussions/2441#discussioncomment-10162225):

However, the fix : `containerRef.current = univer;` causing the problem that `fUniverRef.current.undo()` is no longer working.

```
	useEffect(() => {
		init(data)

		fUniverRef.current.onCommandExecuted(async command => {
			if (command.id === 'sheet.command.set-worksheet-name') {
				console.log('>>> set worksheet name')
				fUniverRef.current.undo()
			}
		})

		return () => {
			destroyUniver()
		}
	}, [data])
```

Can someone give some help on this please ?

**Comment by jikkai** [comment, marked as answer]
There could be two possible reasons for this issue. 

The first one might be the use of React's `<StrictMode>`, which Univer's DI does not currently support, so you need to remove it.
The second reason could be due to an incorrect upgrade of `@wendellhu/redi` during dependency upgrades, which is a peerDependency, resulting in multiple different versions of redi appearing in the project. We have built-in this dependency into `@univerjs/core` in the latest 0.2.5 version, so this problem will no longer occur in future upgrades.
